### PR TITLE
Advanced File Output

### PR DESCRIPTION
**Pull request by pyciko**
This type of file output is mainly targeted for low-cost IoT platforms (like RPi). It optimises some of the aspects of the original File Output, specifically:
- It has an internal temporary buffer which accumulates the logs and prints them out after a certain period. This reduces the frequency of file writes which is good for microSD storages and other types of cheap storages on IoT devices. It is possible to specify the levels of messages which are printed out immediately, without waiting for the next timer tick. By default these are warnings, errors and fatals.
- It has a simple manager for dynamically creating log files instead of using a single one specified by the user. The log files are created in two scenarios: on logger init and when the current log file reaches a size limit. The file names consists of a date+time followed by either "-init" or "-next" suffix depending on the reason why the file was created. This is useful for writing "archives" of telemetry data and logs while keeping them structured.

### CODE REVIEW

**Comment by Bungeefan** [comment]
Hi! Thanks for the nice pull request!

However, before further reviewing your request, unfortunately, the check for Dart 2.17 failed.
`Platform.lineTerminator` was only introduced in [Dart 3.1.0](https://github.com/dart-lang/sdk/blob/3.1.0/CHANGELOG.md#dartio) and this library tries to maintain compatibility with at least 2.17.

I would suggest that you simply replace your usages with the original implementation behind the property:
```dart
static String get lineTerminator => isWindows ? '\r\n' : '\n';
```
https://github.com/dart-lang/sdk/blob/3.1.0/sdk/lib/io/platform.dart#L245

**Comment by pyciko** [comment]
Done

File: lib/src/outputs/advanced_file_output.dart [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+    await _sink?.flush();
+    await _sink?.close();
+    _sink = null; //explicitly make null until assigned again
+  }
+
+  @override
+  Future<void> destroy() async {
+    _bufferWriteTimer?.cancel();
+    _targetFileUpdater?.cancel();
+    await _closeCurrentFile();
```
Comment (Bungeefan) [inline lib/src/outputs/advanced_file_output.dart:146]: As `_closeCurrentFile` only flushes the sink, the buffer should be flushed here too.
This should also be considered in the test cases to ensure that no data can be lost.

File: lib/src/outputs/advanced_file_output.dart [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+        // try creating another file and working with it
+        await _closeCurrentFile();
+        await _openFile(proposed);
+      }
+    }
+  }
+
+  Future<void> _openFile(File proposed) async {
+    _targetFile = proposed;
+    _sink = _targetFile?.openWrite(
```
Comment (Bungeefan) [inline lib/src/outputs/advanced_file_output.dart:130]: Small note: Use the Non-null assertion operator (`!`) instead, as it would be an error to not have a valid file at this point.

File: lib/src/outputs/advanced_file_output_stub.dart [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    Encoding encoding = utf8,
+    List<Level>? writeImmediately,
+    Duration maxDelay = const Duration(seconds: 2),
+    int maxBufferSize = 2000,
+    int maxLogFileSizeMB = 1,
+  }) {
+    throw UnsupportedError("Not supported on this platform.");
+  }
+
+  File? get targetFile => null;
```
Comment (Bungeefan) [inline lib/src/outputs/advanced_file_output_stub.dart:23]: Some of your public fields are missing here (means they can't really be accessed by the developer).
Please either make them private or provide proper getters for them. (Note: Getters can throw `UnsupportedErrors` too!)

**Review by Bungeefan — CHANGES_REQUESTED** [review]
Thanks for the nice pull request, please check my comments and feel free to ask if you need more clarification!

**Comment by Bungeefan** [comment]
That was quick, thanks!
I submitted a review (my first ever on GitHub), you should be able to see it now.
Don't be put off by the amount of comments, I still consider your pull request as well-made!

File: lib/src/outputs/advanced_file_output.dart [hunk 4]
```diff
@@ -0,0 +1,147 @@
+import 'dart:async';
+import 'dart:convert';
+import 'dart:io';
+
+import '../log_level.dart';
+import '../log_output.dart';
+import '../output_event.dart';
+
+extension _NumExt on num {
+  String get twoDigits => toString().padLeft(2, '0');
```
[the remark this answers is not in the document]
Comment (pyciko) [reply lib/src/outputs/advanced_file_output.dart:10]: Okay

File: lib/src/outputs/advanced_file_output.dart [hunk 5]
```diff
@@ -0,0 +1,147 @@
...
+  String get threeDigits => toString().padLeft(3, '0');
+}
+
+/// Writes the log output to a file.
+class AdvancedFileOutput extends LogOutput {
+  AdvancedFileOutput({
+    this.directory,
+    this.file,
```
[the remark this answers is not in the document]
Comment (pyciko) [reply lib/src/outputs/advanced_file_output.dart:18]: Okay, yeah, I tried to make it have the same structure as the existing FileOutput. Will change it

File: lib/src/outputs/advanced_file_output.dart [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+  }
+
+  void _writeOutBuffer() {
+    if (_sink == null) return; //wait until _sink becomes available
+    for (final event in _buffer) {
+      _sink?.writeAll(event.lines, Platform.lineTerminator);
+      _sink?.writeln();
+    }
+    _buffer.clear();
+  }
```
[the remark this answers is not in the document]
Comment (pyciko) [reply lib/src/outputs/advanced_file_output.dart:99]: Not sure... Especially for frequent writes (like a stream of errors) this will result in additional overhead of size checking on every write operation. I'd rather stick with a timer solution and allow the files to be slightly bigger than the size limit, rather than doing this on every write operation.

File: lib/src/outputs/advanced_file_output.dart [hunk 7]
```diff
@@ -0,0 +1,147 @@
...
+      try {
+        if (await _targetFile!.length() > maxLogFileSizeMB * 1000000) {
+          await _closeCurrentFile();
+          await _openFile(proposed);
+        }
+      } catch (e) {
+        // try creating another file and working with it
+        await _closeCurrentFile();
+        await _openFile(proposed);
+      }
```
[the remark this answers is not in the document]
Comment (pyciko) [reply lib/src/outputs/advanced_file_output.dart:124]: I don't think so, because _openFile(proposed) is not meant to be called if current file size is smaller than the limit.

File: lib/src/outputs/advanced_file_output.dart [hunk 8]
```diff
@@ -0,0 +1,147 @@
...
+
+  Future<void> _updateTargetFile() async {
+    if (!dynamicFilesMode) {
+      await _openFile(file!);
+      return;
+    }
+
+    final t = DateTime.now();
+    final newName =
+        '${t.year}-${t.month.twoDigits}-${t.day.twoDigits}_${t.hour.twoDigits}-${t.minute.twoDigits}-${t.second.twoDigits}-${t.millisecond.threeDigits}';
```
[the remark this answers is not in the document]
Comment (pyciko) [reply lib/src/outputs/advanced_file_output.dart:109]: I think maybe it's worth making a `String Function(DateTime time) genFileName` to allow users change the default behaviour. At least is seems as an easier option comparing to things like placeholders etc.. What do you think?

File: lib/src/outputs/advanced_file_output.dart (hunk 6 above)
[the remark this answers is not in the document]
Comment (Bungeefan) [reply lib/src/outputs/advanced_file_output.dart:104]: The size check is a good point! I mean, we could still try to track the size by ourselves ;)
Anyway, your approach seems sensible to me then!

File: lib/src/outputs/advanced_file_output.dart (hunk 5 above)
[the remark this answers is not in the document]
Comment (Bungeefan) [reply lib/src/outputs/advanced_file_output.dart]: Thanks! Yeah, no problem, you couldn't really have known about that requirement!

File: lib/src/outputs/advanced_file_output.dart (hunk 8 above)
[the remark this answers is not in the document]
Comment (Bungeefan) [reply lib/src/outputs/advanced_file_output.dart]: Yeah, sure, that's what I had in mind!
However, I would like to change the name of it, but couldn't yet decide on a fitting one:

* `dateFormatter` (a little bit too generic, no hint that it's "only" for the file name)
* `formatFileFormat` (no hint that it's used for the file name, not content, ...)

I settled on `fileNameFormatter` for now, but I am not sure if that is any better?
Do you have an opinion on the parameter name?

File: lib/src/outputs/advanced_file_output.dart (hunk 7 above)
[the remark this answers is not in the document]
Comment (Bungeefan) [reply lib/src/outputs/advanced_file_output.dart:129]: Sorry, I didn't fully phrase what I meant.
Something like this:
```dart
if (await _targetFile!.length() > maxLogFileSizeMB * 1000000) {
  try {
    await _closeCurrentFile();
  } catch (e, s) {
    print(e);
    print(s);
  } finally {
    await _openFile(proposed);
  }
}
```
_I also added logging, otherwise the developer has little chance to troubleshoot the problem (should one arise)._

I was also under the initial assumption that `_closeCurrentFile` was the only real "catch-able" exception (sensible to catch and continue) in this block.
However, I am not sure if `sink.flush()` or `sink.close()` ever really throws, and it would probably make more sense to also guard the `length` check (as you did).

To sum it up, I would just like you to add the logging and guard the second call to `_closeCurrentFile` (as we can actually continue without successfully closing the old file but not without a new file).

[the remark this answers is not in the document]
Comment (pyciko) [reply lib/src/outputs/advanced_file_output.dart:129]: Okay, I'll add logging. About catchable exceptions, _targetFile!.length() can throw if the target file is missing. This can happen if an external user or program deleted it while the app was running. On newer versions of Dart, it throws somethings like FileMissingException (forgot the actual name), but it's not yet available on Dart 2.17, so I replaced it with a general catch clause for all exceptions.

File: lib/src/outputs/advanced_file_output.dart [hunk 9]
```diff
@@ -144,5 +149,6 @@ class AdvancedFileOutput extends LogOutput {
     _bufferWriteTimer?.cancel();
     _targetFileUpdater?.cancel();
     await _closeCurrentFile();
+    _buffer.clear();
```
Comment (Bungeefan) [inline lib/src/outputs/advanced_file_output.dart:152]: This just clears it (deleting the non-flushed data), I meant writing (and clearing) it before ultimately closing the file:
```dart
_bufferWriteTimer?.cancel();
_targetFileUpdater?.cancel();
try {
  _writeOutBuffer();
} catch (e, s) {
  print("Failed to finally flush buffer: $e");
  print(s);
}
await _closeCurrentFile();
```

File: lib/src/outputs/advanced_file_output.dart (hunk 7 above)
[the remark this answers is not in the document]
Comment (Bungeefan) [reply lib/src/outputs/advanced_file_output.dart:129]: Thanks!
Exactly, yeah, don't restrict the type, I don't think any exceptions should be propagated here.

File: lib/src/outputs/advanced_file_output.dart (hunk 9 above)
Comment (pyciko) [reply lib/src/outputs/advanced_file_output.dart:152]: Okay, changed it, and I'll also add a test to send output events and then quickly close the logger.

File: lib/src/outputs/advanced_file_output_stub.dart (hunk 3 above)
Comment (pyciko) [reply lib/src/outputs/advanced_file_output_stub.dart]: Okay. I think I'll make most of them private because changing them after logger creation may result in unexpected behaviour + it's generally hard to image why someone will need getters for these things...

File: lib/src/outputs/advanced_file_output.dart [hunk 10]
```diff
@@ -0,0 +1,147 @@
...
+    }
+
+    final t = DateTime.now();
+    final newName =
+        '${t.year}-${t.month.twoDigits}-${t.day.twoDigits}_${t.hour.twoDigits}-${t.minute.twoDigits}-${t.second.twoDigits}-${t.millisecond.threeDigits}';
+    if (_targetFile == null) {
+      // just create a new file on first boot
+      await _openFile(File('${directory!.path}/${newName}_init.txt'));
+    } else {
+      final proposed = File('${directory!.path}/${newName}_next.txt');
```
[the remark this answers is not in the document]
Comment (pyciko) [reply lib/src/outputs/advanced_file_output.dart]: I pushed a commit with this change, but IMO it's questionable.
If logger wasn't destroyed properly, the latest.log file may not be renamed to the actual timestamp, and then, on the next launch, we'll have to detect it and rename to something like 1234-56...log.lost to avoid overwriting it with a new file. IMO, it's better to avoid renaming and stick with a format where the name is fixed right from the beginning.

[the remark this answers is not in the document]
Comment (Bungeefan) [reply lib/src/outputs/advanced_file_output.dart]: Why not just check if a `latest.log` file exists on startup?
Then you can either reuse it (append) or rename it now if it is greater than max size?
No need to handle any crashes exceptionally.

You had two different file names even before my comment, `init` and `next`, I just tried to make them more familiar for developers.

**Comment by Bungeefan** [comment]
Feel free to overwrite my changes as you see fit.

**Comment by Bungeefan** [comment]
Please review the new documentation, I tried to implement your description as well as I could!

**Comment by pyciko** [comment]
Thanks, yeah, English is not my native lang, your text looks much better.

File: lib/src/outputs/advanced_file_output.dart (hunk 10 above)
[the remark this answers is not in the document]
Comment (pyciko) [reply lib/src/outputs/advanced_file_output.dart]: So you mean that latest.log file should always be appended until it reaches the size limit, even if the program has rebooted many times, it still has to write everything to the same file with the size limit being the only factor that pushes a new file and renames the older one, right?

[the remark this answers is not in the document]
Comment (Bungeefan) [reply lib/src/outputs/advanced_file_output.dart]: Yes, I assumed that was your basic idea. It also closely reassembles something known as a RollingFile handler, which usually behaves similar to this.
Would you like to change something about that approach?

Addendum: You could also add other constraints for the rotation of log files, however, this would significantly increase the complexity of this class and is probably better implemented in a separate output.

File: lib/src/outputs/advanced_file_output.dart [hunk 11]
```diff
@@ -0,0 +1,194 @@
...
+
+      _targetFileUpdater = Timer.periodic(
+        const Duration(minutes: 1),
+        (_) => _updateTargetFile(),
+      );
+    }
+
+    _bufferFlushTimer = Timer.periodic(_maxDelay, (_) => _flushBuffer());
+    await _openSink();
+    await _updateTargetFile(); // Run first check without waiting for timer tick
```
Comment (Bungeefan) [inline lib/src/outputs/advanced_file_output.dart:127]: I would suggest we also skip this call, if `_maxFileSizeKB` is not a valid size.

```suggestion
    if (_rotatingFilesMode) {
      await _updateTargetFile(); // Run first check without waiting for timer tick
    }
```

File: lib/src/outputs/advanced_file_output.dart (hunk 10 above)
[the remark this answers is not in the document]
Comment (Bungeefan) [reply lib/src/outputs/advanced_file_output.dart]: Thanks for the changes! I am fine with the new rotating logic, so I guess we can resolve this now.
How about you?

[the remark this answers is not in the document]
Comment (pyciko) [reply lib/src/outputs/advanced_file_output.dart]: Yep, everything seems alright

**Review by Bungeefan — APPROVED** [review]

**Comment by pyciko** [comment]
Thx for approval, can't wait to see it released!

**Comment by Bungeefan** [comment]
Sure, maybe next time don't close it if you want it to be merged ;)

Anyway, thanks again for the nice FileOutput!
I think it's a fantastic addition to this project, and thank you for keeping up with my repeated change requests!

**Comment by Bungeefan** [comment]
logger [v2.2.0](https://pub.dev/packages/logger/versions/2.2.0) is now available on pub.dev!

**Comment by neiljaywarner** [comment]
so sorry if this is a dumb question, but does this library include file logging or is it a possible future thing?

**Comment by pyciko** [comment]
> so sorry if this is a dumb question, but does this library include file logging or is it a possible future thing?

If you're using version 2.2.0 or newer, it does include Advanced File Logging with functions described above. It also includes basic FileLogger that existed even earlier.
